crates/ruff_cli/src/diagnostics.rs

```diff
@@ -130,6 +132,19 @@ pub(crate) fn lint_path(
 
     debug!("Checking: {}", path.display());
 
+    // We have to special case this here since the python tokenizer doesn't work with toml
+    if is_project_toml(path) {
+        let contents = std::fs::read_to_string(path)?;
+        let source_file = SourceFileBuilder::new(path.to_string_lossy(), contents).finish();
```

Comment:
This `clones` the whole string. I think it would be better only to construct the `SourceFile` when there are any messages. Or we could create a `SourceFileRef` if we want to pass the path and contents as a single unit.